packages/zero-solid/src/solid-view.ts

```diff
@@ -49,56 +50,130 @@ export class SolidView<V> implements Output {
+    try {
+      this.#applyChanges(this.#pendingChanges, c => c);
+    } finally {
+      this.#pendingChanges = [];
+    }
   };
 
   push(change: Change): void {
-    applyChange(this.#root, change, this.#input.getSchema(), '', this.#format);
+    // Delay updating the solid store state until the transaction commit
+    // (because each update of the solid store is quite expensive), but
+    // read the relationships now as they are only valid to read
+    // when the push is received.
+    this.#pendingChanges.push(materializeRelationships(change));
   }
+
+  #applyChanges<T>(changes: Iterable<T>, mapper: (v: T) => ViewChange): void {
+    this.#setState(oldState => {
+      // Optimization: if the store is currently empty build up
+      // the view on a new plain old JS object root, and return that
+      // for the new state.  This avoids building up large views from
+      // scratch via solid produce.  The proxy object used by solid produce
+      // is slow and in this case we don't care about solid tracking the fine
+      // grained changes (everything has changed, its all new).  For a test case
+      // with a view with 3000 rows, each row having 2 children, this
+      // optimization reduced #applyChanges time from 743ms to 133ms.
+      if (
+        oldState[0][''] === undefined ||
+        (Array.isArray(oldState[0]['']) && oldState[0][''].length === 0)
+      ) {
+        const root: Entry = this.#initialEmptyEntry();
+        this.#applyChangesToRoot<T>(changes, mapper, root);
+        return [root, oldState[1]];
+      }
+      return produce((draftState: State) => {
+        this.#applyChangesToRoot<T>(changes, mapper, draftState[0]);
+      })(oldState);
+    });
+  }
+
+  #applyChangesToRoot<T>(
+    changes: Iterable<T>,
+    mapper: (v: T) => ViewChange,
+    root: Entry,
+  ) {
+    for (const change of changes) {
+      applyChange(
+        root,
+        mapper(change),
+        this.#input.getSchema(),
+        '',
+        this.#format,
+      );
+    }
+  }
+
+  #initialEmptyEntry(): Entry {
+    return {
+      '': this.#format.singular ? undefined : [],
+    };
+  }
+}
+
+function materializeRelationships(change: Change): ViewChange {
+  switch (change.type) {
+    case 'add':
+      return {type: 'add', node: materializeNodesRelationships(change.node)};
+    case 'remove':
+      return {type: 'remove', node: materializeNodesRelationships(change.node)};
+    case 'child':
+      return {
+        type: 'child',
+        node: {row: change.node.row},
+        child: {
+          relationshipName: change.child.relationshipName,
+          change: materializeRelationships(change.child.change),
+        },
+      };
+    case 'edit':
+      return {
+        type: 'edit',
+        node: {row: change.node.row},
+        oldNode: {row: change.oldNode.row},
+      };
+  }
+}
+
+function materializeNodesRelationships(node: Node): Node {
+  return {
+    row: node.row,
+    relationships: Object.fromEntries(
```

Comment:
Do we really need all these arrays?

I'm worried about the strain we are putting on the GC here